### PR TITLE
Add support for "upper bound" variance broadcasting

### PR DESCRIPTION
**Pull request by SimonHeybrock**
This is a preview. Needs more work (verification on paper, as well as more test and better support/docs in code).

Closes scipp/ess#202.
Closes scipp/ess#174.

### CODE REVIEW

File: tests/sans2d_reduction_test.py [hunk 1]
```diff
@@ -66,8 +68,18 @@ def test_can_create_pipeline():
...
+    result = pipeline.compute(BackgroundSubtractedIofQ)
+    assert result.dims == ('Q',)
+
+
+def test_pipeline_can_compute_background_subtracted_IofQ_with_dropped_norm_errors():
+    params = make_params()
+    params[UncertaintyBroadcastMode] = UncertaintyBroadcastMode.drop
+    pipeline = sciline.Pipeline(sans2d_providers(), params=params)
     result = pipeline.compute(BackgroundSubtractedIofQ)
     assert result.dims == ('Q',)
```
Comment (nvaytet) [inline tests/sans2d_reduction_test.py:84]: Is it worth having a check that the variances on the result are smaller than in the case of `UncertaintyBroadcastMode.upper_bound`?

**Comment by nvaytet** [comment]
Will this PR effectively close https://github.com/scipp/ess/issues/202?

File: docs/examples/sans2d.ipynb [hunk 2]
```diff
@@ -208,6 +232,47 @@
...
+    "Mode = NewType('Mode', str)\n",
+    "param_table = sciline.ParamTable(\n",
+    "    Mode,\n",
+    "    {\n",
+    "        UncertaintyBroadcastMode: [\n",
+    "            UncertaintyBroadcastMode.upper_bound,\n",
+    "            UncertaintyBroadcastMode.drop,\n",
+    "        ]\n",
+    "    },\n",
+    "    index=[Mode('upper_bound'), Mode('drop')],\n",
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply docs/examples/sans2d.ipynb:268]: In fact I am thinking about supporting this better in Sciline. The current interface seems tailored towards having a table with multiple params columns where the above makes more sense. But here we actually just want a list of params.

File: src/esssans/i_of_q.py [hunk 3]
```diff
@@ -72,21 +78,16 @@ def preprocess_monitor_data(
...
-                'contributions to uncertainties from correlations.'
+        if uncertainties == UncertaintyBroadcastMode.drop:
+            monitor -= sc.values(background)
+        elif uncertainties == UncertaintyBroadcastMode.upper_bound:
+            monitor -= broadcast_with_upper_bound_variances(
+                background, sizes=monitor.sizes
             )
-        monitor = monitor - bg
+        else:
+            monitor -= background
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/esssans/i_of_q.py:90]: Yes, there is no reason to raise ourselves: If someone calls it on data with variances this will fail here and give the usual error message.

File: src/esssans/normalization.py [hunk 4]
```diff
@@ -160,38 +130,40 @@ def iofq_denominator(
...
-    #    _verify_normalization_alpha(
-    #        numerator=data.hist(wavelength=denominator.coords['wavelength']),
-    #        denominator=denominator,
-    #        signal_over_monitor_threshold=signal_over_monitor_threshold,
-    #    )
-
-    denominator = solid_angle * sc.values(denominator)
+    if uncertainties == UncertaintyBroadcastMode.drop:
+        if direct_beam is not None:
+            denominator = direct_beam * denominator
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/esssans/normalization.py:152]: Good catch, I need to use `sc.values` here as well!

File: src/esssans/normalization.py [hunk 5]
```diff
@@ -221,6 +193,10 @@ def normalize(
     :
         The input data normalized by the supplied denominator.
     """
+    if denominator.variances is not None and numerator.bins is not None:
+        # Event-mode normalization is not correct of norm-term has variances.
+        # See https://doi.org/10.3233/JNR-220049 for context.
+        numerator = numerator.hist()
     if numerator.bins is not None:
         da = numerator.bins / sc.lookup(func=denominator, dim='Q')
     else:
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/esssans/normalization.py:202]: I think this is too be discussed with Wojciech. Maybe they never want events, then we can get rid of these branches.

File: src/esssans/types.py [hunk 6]
```diff
@@ -31,10 +32,19 @@
 Denominator = NewType('Denominator', sc.DataArray)
 """Denominator of IofQ"""
 IofQPart = TypeVar('IofQPart', Numerator, Denominator)
-"""TypeVar used for specifying Numerator or Denominator of IofQ""" ""
+"""TypeVar used for specifying Numerator or Denominator of IofQ"""
 
 # 2  Workflow parameters
 
+UncertaintyBroadcastMode = Enum(
+    'UncertaintyBroadcastMode', ['drop', 'upper_bound', 'fail']
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/esssans/types.py:40]: If there are variances, broadcasting will fail, since Scipp raises.

File: src/esssans/uncertainty.py [hunk 7]
```diff
@@ -2,37 +2,36 @@
...
+def broadcast_with_upper_bound_variances(
+    data: Union[sc.Variable, sc.DataArray], sizes: Dict[str, int]
+) -> Union[sc.Variable, sc.DataArray]:
+    if all(data.sizes.get(dim) == size for dim, size in sizes.items()):
+        return data
+    size = 1
+    for dim, dim_size in sizes.items():
+        if dim not in data.dims:
+            size *= dim_size
+    data = data.copy()
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/esssans/uncertainty.py:35]: On the contrary, it is very *relevant* to do it in this order: This copy is tiny. With the current implementation we have:

```
data = data.copy()  # 1 read + 1 write
data.variances = data.variances * size  # 1 read + 1 write
data.broadcast().copy()  # 1 read + N write
```
With what you suggest you would have:
```
data = data.broadcast().copy()  # 1 read + N write
data.variances *= size  # N read + N write
```
so the required memory bandwidth is about 3x worse. Not counting the cost for the large allocation (which may dominate) you would cut performance to 1/3.

File: src/esssans/uncertainty.py [hunk 8]
```diff
@@ -2,37 +2,36 @@
...
+def broadcast_with_upper_bound_variances(
+    data: sc.DataArray, sizes: Dict[str, int]
+) -> sc.DataArray:
+    pass
+
+
+def broadcast_with_upper_bound_variances(
+    data: Union[sc.Variable, sc.DataArray], sizes: Dict[str, int]
+) -> Union[sc.Variable, sc.DataArray]:
+    if all(data.sizes.get(dim) == size for dim, size in sizes.items()):
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply src/esssans/uncertainty.py:29]: There is probably a better API for this, maybe by passing in all the operands that will take part in an operation? We can adapt it in the future once we see this used in more places, to get a better feeling about the concrete requirements and usability.

**Comment by SimonHeybrock** [comment]
> Will this PR effectively close [scipp/ess#202](https://github.com/scipp/ess/issues/202)?

Yes, provided that we can avoid backsliding. We should keep on track to sunset `ess.sans` ASAP.

I edited this (and another) issue as fixed in the PR message.

File: tests/sans2d_reduction_test.py (hunk 1 above)
Comment (SimonHeybrock) [reply tests/sans2d_reduction_test.py:84]: Good question, is this something to be tested at this level, or do we want to rely on Sciline to do its thing and instead add tests to this effect on the individual provider level?

File: src/esssans/normalization.py [hunk 9]
```diff
@@ -160,38 +130,42 @@ def iofq_denominator(
...
-    #        numerator=data.hist(wavelength=denominator.coords['wavelength']),
-    #        denominator=denominator,
-    #        signal_over_monitor_threshold=signal_over_monitor_threshold,
-    #    )
-
-    denominator = solid_angle * sc.values(denominator)
+    if uncertainties == UncertaintyBroadcastMode.drop:
+        if direct_beam is not None:
+            denominator = direct_beam * sc.values(denominator)
+        if set(solid_angle.dims) < set(denominator.dims):
```
Comment (nvaytet) [inline src/esssans/normalization.py:153]: In our sans2d workflow, the solid angle has dimension of `pixel`, and the denominator has dimension `wavelength`. I think in that case the condition comparing the sets is always `False`, so you are not stripping the variances?

Unless I misunderstood something?

File: src/esssans/normalization.py [hunk 10]
```diff
@@ -160,38 +130,42 @@ def iofq_denominator(
...
-    #    _verify_normalization_alpha(
-    #        numerator=data.hist(wavelength=denominator.coords['wavelength']),
-    #        denominator=denominator,
-    #        signal_over_monitor_threshold=signal_over_monitor_threshold,
-    #    )
-
-    denominator = solid_angle * sc.values(denominator)
+    if uncertainties == UncertaintyBroadcastMode.drop:
+        if direct_beam is not None:
+            denominator = direct_beam * sc.values(denominator)
```
Comment (nvaytet) [inline src/esssans/normalization.py:152]: How do we know, in a general case, that we have to strip the variances of the `denominator` and not the ones of the `direct_beam`?
Maybe in this particular example, it doesn't matter because they get stripped again below before multiplying by the solid angle.

I guess then the assumption that is made is that the `solid_angle` has no variances, because if it does, then it might not be obvious that we need to strip variances from denominator and not the ones from the solid angle.
Basically, should we strip variances from both operands in the broadcast, instead of just one of them?

File: src/esssans/normalization.py (hunk 9 above)
Comment (SimonHeybrock) [reply src/esssans/normalization.py:153]: Yes, I wrote this in a confusing manner... the monitor errors were dropped above, so this should only be entered when the direct beam is pixel dependent. I'll rewrite.

File: src/esssans/normalization.py (hunk 10 above)
Comment (SimonHeybrock) [reply src/esssans/normalization.py:152]: > How do we know, in a general case, that we have to strip the variances of the denominator and not the ones of the direct_beam?

Because the direct beam is wavelength dependent and is never broadcasted here?

File: src/esssans/normalization.py (hunk 9 above)
Comment (SimonHeybrock) [reply src/esssans/normalization.py]: Updated, please have another look.
